course-content/13_working_with_code.qmd

```diff
@@ -0,0 +1,397 @@
+ <br>
+ 
+<center>
+    <img src="images/working-with-code/code_quality_3.png" size=700 style="margin:auto"/>
+    <p style="text-align: center">
+        Style matters!
+    </p>
+</center>
+
+Fortunately, there are some widely accepted guidelines about coding style that you should be aware of and strive towards over time. In the Python world, the style guide you should know about is "PEP8" which stands for Python Enhancements Proposals. 
+
+There are many specific rules in PEP8, here are some to be aware of from the start. 
+
+*Comments*
+
+When coding, it is good practice to `comment` on code, but only when the code is likely to be non-obvious to others who are likely to review it. Comments are preceded by a "#", which within Python, ensures that they are recognized as non-executable and therefore no threat to actually affecting your code. Below are two examples of how you can add comments to your code (if necessary).
+
+```{python}
+# The bog standard first line of code learned in any progamming language
+print("Hello World")
+
+print("Hello World")  # New programmer's first line of code in any language 
+```
+
+*Spacing within code*
+
+There are a number of little conventions about spacing that just have to be learned. In the code below look at the comments that go along to see what is considered good style and what is considered good and what is not.  
+
+```{python}
+# Good style, as there is a space on either side of the assignment operator 
+a = print("hello")  
+
+# Poor style, as the assignment operator lacks spaces around it
+a=print("hello")    
+```
+
+```{python}
+# Good style, as there is no space between the function call and left parenthesis
+print(2)  
+
+# Poor style, as there should be no space between function call and left parenthesis
+print (2) 
+```
+
+Below are some examples of how to incorporate spaces when you have operators with different priorities (i.e. remember PEMDAS or BODMAS from high school?). In general, add space for those instances with lower priority:
+
+```python
+# Good style
+i = i + 1 
+x = x*2 - 1
+c = (a+b) * (a-b)
+
+# Poor style
+i=i+1 
+x = x * 2 - 1
+c = (a + b) * (a - b)
+```
+
+There are many other conventions about spacing, but the ones above are some of the more common ones to watch out for. 
+
+
+*Indentation*
+
+A line of code that completes a meaningful coherent action of some sort, is kind of like a complete sentence in grammar. As with sentences, a line of code can be short and sweet, long and convoluted, and anywhere in between. Where lines of code are longer than 79 characters, certain style guide rules come into play. The examples below show how to carry over content from one line to the next line. 
+
+```python
+# Correct - aligned with opening delimeter
+animals = animals.drop('weight', 'size', 'age', 'location', 'eye colour', 
+                       'intelligence', 'cost', 'number held in zoos')
+
+# Correct - hanging indents (usually four spaces, should not use tabs)
```

Comment:
I'm totally a tabber 😬. Forgive me pep8!

Reply:
i think it is the mixing of spacing and tabbing that PEP8 "takes a dim view of" ... i am going to rephrase to just use one or the other